src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_IncrementalAnalyzer.cs

```diff
@@ -44,33 +45,58 @@ private async Task AnalyzeDocumentForKindAsync(Document document, AnalysisKind k
-                var compilation = await GetOrCreateCompilationWithAnalyzersAsync(document.Project, stateSets, cancellationToken).ConfigureAwait(false);
+                var compilationWithAnalyzers = await GetOrCreateCompilationWithAnalyzersAsync(document.Project, stateSets, cancellationToken).ConfigureAwait(false);
 
+                // First attempt to fetch diagnostics from the cache, while computing the state sets for analyzers that are not cached.
+                using var _ = ArrayBuilder<StateSet>.GetInstance(out var nonCachedStateSets);
                 foreach (var stateSet in stateSets)
                 {
-                    var analyzer = stateSet.Analyzer;
-
-                    var result = await GetDocumentAnalysisDataAsync(compilation, document, stateSet, kind, cancellationToken).ConfigureAwait(false);
-                    if (result.FromCache)
+                    var data = await TryGetCachedDocumentAnalysisDataAsync(document, stateSet, kind, cancellationToken).ConfigureAwait(false);
+                    if (data.HasValue)
                     {
-                        RaiseDocumentDiagnosticsIfNeeded(document, stateSet, kind, result.Items);
-                        continue;
+                        // We need to persist and raise diagnostics for suppressed analyzer.
+                        PersistAndRaiseDiagnosticsIfNeeded(data.Value, stateSet);
+                    }
+                    else
+                    {
+                        nonCachedStateSets.Add(stateSet);
                     }
+                }
 
-                    // no cancellation after this point.
-                    var state = stateSet.GetOrCreateActiveFileState(document.Id);
-                    state.Save(kind, result.ToPersistData());
+                // Then, compute the diagnostics for non-cached state sets.
+                if (nonCachedStateSets.Count > 0)
+                {
+                    var analysisScope = new DocumentAnalysisScope(document, span: null, nonCachedStateSets.SelectAsArray(s => s.Analyzer), kind);
+                    var executor = new DocumentAnalysisExecutor(analysisScope, compilationWithAnalyzers, DiagnosticAnalyzerInfoCache);
+                    foreach (var stateSet in nonCachedStateSets)
+                    {
+                        var computedData = await ComputeDocumentAnalysisDataAsync(executor, stateSet, cancellationToken).ConfigureAwait(false);
+                        PersistAndRaiseDiagnosticsIfNeeded(computedData, stateSet);
+                    }
 
-                    RaiseDocumentDiagnosticsIfNeeded(document, stateSet, kind, result.OldItems, result.Items);
+                    var asyncToken = AnalyzerService.Listener.BeginAsyncOperation(nameof(AnalyzeDocumentForKindAsync));
+                    var _2 = ReportAnalyzerPerformanceAsync(document, compilationWithAnalyzers, cancellationToken).CompletesAsyncOperation(asyncToken);
```

Comment:
>_2 [](start = 24, length = 2)

nit: `_ = ReportAnalyzerPerformanceAsync(...);`?

Reply:
There is already a discard variable `_` defined in scope earlier in the method.

Reply:
`_ = lhs;` does not define a variable.